aiidalab/app.py

```diff
@@ -26,516 +38,298 @@ class AppNotInstalledException(Exception):
-                if key.endswith(b'HEAD') or key.startswith(b'refs/heads/'):
-                    continue
-                obj = self.repo.get_object(value)
-                if isinstance(obj, Tag):
-                    self._refs_dict[key] = obj.object[1]
-                elif isinstance(obj, Commit):
-                    self._refs_dict[key] = value
-        return self._refs_dict
-
-    @property
-    def available_versions(self):
-        """Function that looks for all the available branches. The branches can be both
-        local and remote.
-
-        : return : an OrderedDict that contains all available branches, for example
-                   OrderedDict([('master', 'refs/remotes/origin/master')])."""
-
-        if not hasattr(self, '_available_versions'):
-            # HEAD branch won't be included
-            if not self.refs_dict:  # if no branches were found - return None
-                return {}
-
-            # Add remote branches.
-            available = OrderedDict({
-                name.split(b'/')[-1].decode("utf-8"): name
-                for name, _ in self.refs_dict.items()
-                if name.startswith(b'refs/remotes/')
-            })
-
-            # Add local branches that do not have tracked remotes.
-            for name in self.refs_dict:
-                if name.startswith(b'refs/heads/'):
-                    branch_label = name.replace(b'refs/heads/', b'').decode("utf-8")
-                    pattern = re.compile("refs/remotes/.*/{}".format(branch_label))
-                    # check if no tracked remotes that correspond to the current local branch
-                    if not any(pattern.match(value) for value in available.values()):
-                        available[branch_label] = name
-
-            # Add tags.
-            available.update(
-                sorted({
-                    name.split(b'/')[-1].decode("utf-8"): name
-                    for name, _ in self.refs_dict.items()
-                    if name.startswith(b'refs/tags/')
-                }.items(),
-                       reverse=True))
-            self._available_versions = available
-        return self._available_versions
-
-    @property
-    def current_version(self):
-        """Function that returns the reference to the currently selected branch,
-        for example 'refs/remotes/origin/master'."""
-
-        if not hasattr(self, '_current_version'):
-
-            # If no branches were found - return None
-            if not self.refs_dict:
-                return None
-
-            # Get the current version
-            available = self.available_versions
+        with self._show_busy():
+            self._stop_watch_repository()
             try:
+                shutil.rmtree(self.path)
+            except FileNotFoundError:
+                raise RuntimeError("App was already uninstalled!")
+            self.refresh()
+            self._watch_repository()
```

Comment:
Just a bit confusing: why do we need to watch the repo if the folder is gone already? I think only local filesystem events are watched, no?

Reply:
Because we need to detect whether the app is installed again. We are watching the parent folder as well. I could solve that more elegantly, but the current solution works.

Reply:
ok.